File: realm/src/main/java/hirez/realm/RealmRoyale.java
package hirez.realm;

import hirez.api.*;
import hirez.api.object.*;
import hirez.api.object.interfaces.Queue;
import hirez.api.object.interfaces.ReturnedMessage;
import hirez.realm.object.*;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Single;
import lombok.extern.slf4j.Slf4j;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.time.ZoneOffset;
import java.util.Arrays;
import java.util.Date;
import java.util.TimeZone;
import java.util.function.Consumer;

@Slf4j
public class RealmRoyale extends Endpoint {
    private static final BaseEndpoint DEFAULT_BASE = new BaseEndpoint() {
        @Override
        public Game getGame() {
            return new Game("35lh77mxwjy9", "Realm Royale");
        }

        @Override
        public Platform getPlatform() {
            return new Platform("4bsbxhr3mrlz", "PC");
        }

        @Override
        public String getBaseUrl() {
            return "https://api.realmroyale.com/realmapi.svc";
        }
    };

    private RealmRoyale(Configuration configuration) {
        super(configuration);
    }

    public static RealmRoyale create(Consumer<ConfigurationBuilder> configuration) {
        return new RealmRoyale(new ConfigurationBuilder().applyFrom((cfg) -> {
            configuration.accept(cfg);
            if (cfg.getBaseEndpoint() == null) {
                cfg.setBaseEndpoint(DEFAULT_BASE);
            }
        }).build());
    }

    public Flowable<Talent> getTalents(Language language) {
        return testAndCall(Talent[].class, "gettalents", language.getId().toString())
                .flattenAsFlowable(Arrays::asList);
    }

    public Single<Leaderboard> getLeaderboard(Queue queue, Criteria criteria) {
        return testAndCall(Leaderboard.class, "getleaderboard", queue.getId().toString(), criteria.id.toString());
    }

    public Single<MatchDetails> getMatchDetails(long matchId) {
        return testAndCall(MatchDetails.class, "getmatchdetails", "17327425");
    }

    public Flowable<MatchId> getMatchIdsByQueue(Queue queue, Date timestamp) {
        DateFormat df = new SimpleDateFormat("yyyyMMdd/hh");
        DateFormat mdf = new SimpleDateFormat("mm");

        int minutes = Integer.parseInt(mdf.format(timestamp));

        minutes = minutes - minutes % 10 + ((minutes % 10) > 4 ? 10 : 0);

        return testAndCall(MatchId[].class, "getmatchidsbyqueue", queue.getId().toString(), df.format(timestamp) + minutes)
                .flattenAsFlowable(Arrays::asList);
    }

    public Single<PlayerMatchHistory> getPlayerMatchHistory(long userId) {
        return testAndCall(PlayerMatchHistory.class, "getplayermatchhistory", Long.toString(userId));
    }

    public Single<PlayerMatchHistory> getPlayerMatchHistoryAfterDateTime(long userId, Date after) {
        DateFormat df = new SimpleDateFormat("yyyyMMddHHmmss");
        df.setTimeZone(TimeZone.getTimeZone(ZoneOffset.UTC));
        return testAndCall(PlayerMatchHistory.class, "getplayermatchhistoryafterdatetime", df.format(after), Long.toString(userId));
    }

    public Single<Player> getPlayer(long userId) {
        return testAndCall(Player.class, "getplayer", Long.toString(userId), "hirez");
    }

    public Single<Player> getPlayerBySteamId(long steamId) {
        return testAndCall(Player.class, "getplayer", Long.toString(steamId), "steam");
    }

    public Flowable<PlayerIdPortal> getPlayerIdByName(String query) {
        return testAndCall(PlayerIdPortal.class, "getplayeridbyname", query)
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<PlayerIdPortal> getPlayerIdByPortalUserId(Portal portal, long id) {
        return testAndCall(PlayerIdPortal[].class, "getplayeridbyportaluserid", portal.getId().toString(), Long.toString(id))
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<PlayerIdPortal> getPlayerIdsByGamerTag(Portal portal, String query) {
        return testAndCall(PlayerIdPortal[].class, "getplayeridsbygamertag", portal.getId().toString(), query)
                .flattenAsFlowable(Arrays::asList);
    }

    public Single<PlayerStats> getPlayerStats(long userId) {
        return testAndCall(PlayerStats.class, "getplayerstats", Long.toString(userId));
    }

    public Single<PlayerStatus> getPlayerStatus(long userId) {
        return testAndCall(PlayerStatus.class, "getplayerstatus", Long.toString(userId));
    }

    public Flowable<PlayerQuery> searchPlayer(String query) {
        return testAndCall(PlayerQuery[].class, "searchplayers", query)
                .flattenAsFlowable(Arrays::asList);
    }
}

File: realm/src/main/kotlin/hirez/realm/RealmRoyale.kt
package hirez.realm

import com.google.gson.Gson
import hirez.Config
import hirez.ConfigBuilder
import hirez.Language
import hirez.api.AbstractAPI
import hirez.api.BaseEndpoint
import hirez.api.Game
import hirez.api.Http
import hirez.api.Platform
import hirez.api.Queue
import hirez.enums.Portal
import hirez.format
import hirez.json.CreateSession
import hirez.json.DataUsage
import hirez.json.Patch
import hirez.json.Ping
import hirez.json.PlayerData
import hirez.json.PlayerState
import hirez.json.QueueMatch
import hirez.json.Server
import hirez.json.TestSession
import hirez.realm.json.Match
import hirez.realm.json.MatchHistory
import hirez.realm.json.Player
import hirez.realm.json.PlayerItem
import hirez.realm.json.PlayerStat
import hirez.realm.json.QueueLeaderboard
import hirez.realm.json.Talent
import hirez.statuspage.StatusPage
import java.util.*

/**
 *
 * @author Damian Staszewski [damian@stachuofficial.tv]
 * @version %I%, %G%
 * @since 1.0
 */
object RealmRoyale : AbstractAPI() {
	@JvmStatic
	fun initConfig(config: Config) {
		this.config = reconfigure(config)
		this.http = Http(Gson(), config.userAgent)
	}
	
	private fun reconfigure(config: Config): Config =
				ConfigBuilder().apply {
					setDevId(config.devId)
					setAuthKey(config.authKey)
					setDefaultLanguage(config.defaultLanguage)
					setSessionStorage(config.sessionStorage)
					setUserAgent(config.userAgent)
					setEndpoint(object : BaseEndpoint {
						override val game: Game = Game("35lh77mxwjy9", "Realm Royale")
						override val platform: Platform = Platform("4bsbxhr3mrlz", "PC")
						override val baseUrl: String = "https://api.realmroyale.com/realmapi.svc"
					})
				}.build()
	
	@JvmStatic
	fun initConfig(configBuilder: ConfigBuilder.() -> Unit) =
				initConfig(ConfigBuilder().apply(configBuilder).build())
	
	@JvmStatic
	fun createsession() = get<CreateSession>("createsession").doOnSuccess(config.sessionStorage::set)
	val dataused
		@JvmStatic get() = get<DataUsage>("getdataused")
	val hirezserverstatus
		@JvmStatic get() = get<Server>("gethirezserverstatus")
	val patchinfo
		@JvmStatic get() = get<Patch>("getpatchinfo")
	
	@JvmStatic
	fun ping() = get<String>("ping").map(::Ping)
	
	@JvmStatic
	fun testsession() = get<String>("testsession")
				.map(::TestSession)
				.doOnSuccess { if (!it.isSuccessful) config.sessionStorage.remove() }
	
	@JvmStatic
	fun getTalents(language: Language) = get<Array<Talent>>("gettalents", language.id.toString())
	val talents
	
	@JvmStatic get() = getTalents(config.defaultLanguage)
	
	@JvmStatic
	fun getLeaderboard(queue: Queue, rankCriteria: RankCriteria) = get<QueueLeaderboard>("getleaderboard", queue.id.toString(), rankCriteria.id.toString())
	@JvmStatic
	fun getMatchDetails(id: Long) = get<Match>("getmatchdetails", id.toString())
	@JvmStatic
	fun getMatchIdsByQueue(queue: Queue, timestamp: Calendar) =
				get<Array<QueueMatch>>("getmatchidsbyqueue", timestamp.apply { set(Calendar.MINUTE, get(Calendar.MINUTE).let { it - (it % 10) }) }.format("yyyyMMdd/HHmm"))
	
	@JvmStatic
	fun getPlayerMatchHistory(playerId: Long) = get<MatchHistory>("getplayermatchhistory", playerId.toString())
	@JvmStatic
	fun getPlayerMatchHistoryAfterDateTime(playerId: Long, timestamp: Calendar) =
				get<MatchHistory>("getplayermatchhistoryafterdatetime", timestamp.format("yyyyMMddHHmmss"))
	
	@JvmStatic
	fun getPlayerSteam(steamId: Long) = get<Player>("getplayer", steamId.toString(), "steam")
	@JvmStatic
	fun getPlayerHirez(id: Long) = get<Player>("getplayer", id.toString(), "hirez")
	@JvmStatic
	fun getPlayerIdByName(name: String) = get<PlayerData>("getplayeridbyname", name)
	@JvmStatic
	fun getPlayerIdByPortalUserId(portal: Portal, id: Long) = get<PlayerData>("getplayeridbyportaluserid", portal.id.toString(), id.toString())
	@JvmStatic
	fun getPlayerIdsByGamertag(portal: Portal, gamertag: String) = get<Array<PlayerData>>("getplayeridsbygamertag", portal.id.toString(), gamertag)
	@JvmStatic
	fun getPlayerStats(id: Long) = get<PlayerStat>("getplayerstats", id.toString())
	@JvmStatic
	fun getPlayerStatus(id: Long) = get<PlayerState>("getplayerstatus", id.toString())
	@JvmStatic
	fun searchPlayers(query: String) = get<Array<PlayerItem>>("searchplayers", query)
	
	@JvmStatic
	val statusPage
		@JvmStatic
		get() = StatusPage(http)
}